### PR TITLE
Use TenantGeneratedCertificatedProxy in queue cretificates module.

### PR DESCRIPTION
**Pull request by andrey-canon**
## Description
This changes overrides the save method for GeneratedCertificate model and allows to use the TenantGeneratedCertificatedProxy in the  lms.djangoapps.certificates.queue module.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -10,10 +10,11 @@ configparser==3.5.0       # via pylint
 coverage==4.5.1
 django-crum==0.7.3
 django-fake-model==0.1.4
+django-model-utils==3.1.2
```
Comment (morenol) [inline requirements.txt]: Is this only for testing purposes?

**Comment by morenol** [comment]
After this, I guess we are going to revert this https://github.com/eduNEXT/edunext-platform/commit/7425630cce7be1dfe2790c4c11300a84ae8a8ea7, can you run the tests there?

File: eox_tenant/tenant_wise/__init__.py [hunk 2]
```diff
@@ -14,6 +14,12 @@ def load_tenant_wise_overrides():
...
-        get_certificates_models().GeneratedCertificate = TenantGeneratedCertificateProxy
+    if getattr(settings, 'USE_EOX_TENANT', False):
+        allowed_proxies = getattr(settings, 'TENANT_WISE_ALLOWED_PROXIES', {})
+
+        if allowed_proxies.get('TenantSiteConfigProxy'):
+            get_site_configuration_models().SiteConfiguration = TenantSiteConfigProxy
+
+        if allowed_proxies.get('TenantGeneratedCertificateProxy'):
+            get_certificates_models().GeneratedCertificate = TenantGeneratedCertificateProxy
+            get_certificates_queue().GeneratedCertificate = TenantGeneratedCertificateProxy
```
Comment (felipemontoya) [inline eox_tenant/tenant_wise/__init__.py]:
```suggestion
            my_proxy_installer('lms.certificates.queue.GeneratedCertificate', TenantGeneratedCertificateProxy)
```

File: eox_tenant/tenant_wise/proxies.py [hunk 3]
```diff
@@ -312,3 +315,28 @@ class Meta:
...
+
+    def save(self, *args, **kwargs):
+        """
+        We need to override the save method in order to call the
+        <lms.djangoapps.certificates.models.GeneratedCertificate> parent method and since we have overridden the
+        <lms.djangoapps.certificates.models.GeneratedCertificate> with
+        <eox_tenant.tenant_wise.TenantGeneratedCertificateProxy> we have to avoid the recurrent calls by the
+        grandparent method.
+        """
+        super(GeneratedCertificate, self).save(*args, **kwargs)  # pylint: disable=bad-super-call
```
Comment (felipemontoya) [inline eox_tenant/tenant_wise/proxies.py]:
```suggestion
        with MyContextProxy('lms.certificates.queue.GeneratedCertificate', NoProxyCert, ProxyCert):
            super(GeneratedCertificate, self).save(*args, **kwargs)  # pylint: disable=bad-super-call
```

File: eox_tenant/tenant_wise/context_managers.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+from contextlib import contextmanager
+
+logger = logging.getLogger(__name__)
+
+
+@contextmanager
+def proxy_regression(module, model, regressive_model):
+    """
+    Allow to execute an operation with a different model to given module.
```
Comment (felipemontoya) [inline eox_tenant/tenant_wise/context_managers.py]:
```suggestion
    Allow to execute an operation with a different model to a given module.
```

File: eox_tenant/tenant_wise/context_managers.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    try:
+        previous_model = getattr(module, model)
+        setattr(module, model, regressive_model)
+        yield
+    except AttributeError as error:
+        logger.warning('The error %s has been generated for the module %s and model %s', error, module, model)
```
Comment (felipemontoya) [inline eox_tenant/tenant_wise/context_managers.py]: I think this is enough to warrant an Error

File: eox_tenant/tenant_wise/proxies.py [hunk 6]
```diff
@@ -312,3 +315,10 @@ class Meta:
     def __unicode__(self):
         key = getattr(settings, "EDNX_TENANT_KEY", "No tenant is active at the moment")
         return u"<Tenant proxy as GeneratedCertificate: {}>".format(key)
+
+    def save(self, *args, **kwargs):
+        """
+        Override save method. This allows to use the super save method.
```
Comment (felipemontoya) [inline eox_tenant/tenant_wise/proxies.py]:
```suggestion
        Override the save method such that we use the non proxy cert during the save.
```

File: eox_tenant/tenant_wise/__init__.py [hunk 7]
```diff
@@ -3,17 +3,48 @@
...
+            set_as_proxy(
+                modules=certificate_list,
+                model='GeneratedCertificate',
+                proxy=TenantGeneratedCertificateProxy
+            )
+
+
+def set_as_proxy(modules, model, proxy):
+    """
+    Set a model as a proxy.
```
Comment (felipemontoya) [inline eox_tenant/tenant_wise/__init__.py]:
```suggestion
    Helper to patch a loaded module with a proxy object that has all the Tenant wise properties.
```

**Pull request merged**